### DISCUSSION TITLE
Discussion (Portable Side)

### DISCUSSION THREAD

**Discussion by nachotacos69** [HaoJun0823/GECV#24] [General]
Discussion on the Portable sides. Leaving this here to avoid chattering inside the issues

**Comment by nachotacos69** [comment]
### Explanation on SPRX files containing texts


Though, I'm not some kind of expert on things, here and there. But I'm gonna leave my findings here.

-  SPRX (Signed Playstation Relocatable eXecutable) are some kind of libraries. But I'm not gonna dig into it much since im not good at digging ELF files and doing disassembling/reassembling isn't my style.


♪ Translation Sample: `god_g998_Results.sprx`
![snap002](https://github.com/user-attachments/assets/8071ca63-1b52-48cf-a1c8-2a4280d2cc0d)


- Some SPRX (or mostly) have texts within that are outside of the `kst`, `tr2`, files. Thus making them hard to translate part as the text are limited when replacing them. This is a problem when translating certain parts of the game that have lengthy text after the Japanese text translated to English or other languages. So some might just go break out of nowhere and crashes the game when it shows (or doesn't but ignores that overlapped texts). Take for example: `Survival Profile` doesn't exists inside the result `.tr2` file (same goes for other texts that you see except the npc/player name).

**Reply by nachotacos69** [reply]
Some more image samples

♪ same stuff from `god_g998_Results.sprx`
![snap001](https://github.com/user-attachments/assets/e0bf32f7-9bdc-4c44-83b1-c946aa53e234)

♪ inside of `god_m001_Title.sprx`
![NPJH50832_00012](https://github.com/user-attachments/assets/afab1738-71d5-4e55-8730-d6720091c9df)

**Comment by DarkGD123** [comment]
hmm, how bout those scattered radio chatter texts?

**Reply by nachotacos69** [reply]
stored within some `String_str` files?

**Comment by nachotacos69** [comment]
## EBOOT

EBOOT is like your PC's EXE file, but different. it's a `ELF32` Little Endian MIPS.
![ELF32](https://github.com/user-attachments/assets/1c49963a-c9fe-4e0e-8b89-0072007849ff)



- This controls most of the operations here and there when the game boots and certain stuff here and there. it also controls the text limitations of the characters bytes

**Reply by nachotacos69** [reply]
- Sample Image 1 (using 1.30): if text/character limitation is not followed. for example for both Mission Name/Enemy Name (space included for each text). especially with the enemy name with `10` character limitation text.

![NPJH50832_00017](https://github.com/user-attachments/assets/3a13ce5a-8ec0-4c71-a686-10471bc7ce5c)
- This will crash on the PSP. but will not crash on PPSSPP but cut's the rest of the text and shows error in the log console

**Comment by nachotacos69** [comment]
## repetitive model usage (on enemies)


but my theories on them using same models and placed on different location is due to their pathways and such (fpth and more)

thus the game size is quite big. so if we find a way to optimize these stuff, we could make the game more lightweight

**Comment by nachotacos69** [comment]
# encrypted DATA.RDP

there's a reason why i put that words on full caps. there's another version of data.rdp named **DATA.RDP** inside of `/PSP_GAME/INSDIR/` (if you view the ISO in 7z or PowerISO)

This is for the **INSTALL GAME DATA** function in the game main screen. **DATA.RDP** file's purpose is for those who play GOD EATER 2 in a UMD discs. since we use ISO, this is unnecessary. its function is still the same as **data.rdp**
![umd data](https://github.com/user-attachments/assets/353bde99-596b-4fa7-a39b-9b4138bfc37c)

it is also encrypted in PGD "Protected Game Data" data formatting.

**Comment by nachotacos69** [comment]
# ATRAC3Plus audio (.AT3/AT3X)
`If you have a PSPSDK (official), you should see a at3tool in there.`

-These audio format is used in environment effects, radio chatter, cutscenes, and background music.

## Information
- MAGIC/Header: RIFF
- Second Magic: file size (UINT32)
- Third Magic: WAVEfmt
- The rest: important chunks and values.
File Sizes Limitations (Tested only for GEB/GE2)
- Size (GEB): 512KB minimal for loop. 5MB minimal for one time only audio (cutscenes)
- Size (GE2): 1.0MB minimal for loop?. And 8MB minimal one time audio? (radio chatter/cutscenes/etc.)
- Audio Quality: can handle 48KHz and 64KHz and above but look out for the size.

## Looping
- Looping audio takes time, since you need an actual part of an audio to be perfectly matched before it reaches to it's end. Take for example, the battle theme of `Rufus Caligula` is perfectly looped. And Sony is very strict when it comes to audio looping in AT3 files (refer to the at3tool to check for the stricts)

**Comment by nachotacos69** [comment]
# ZLIB Compression
- https://github.com/HaoJun0823/GECV-OLD/blob/master/GECV_EX_BLZ4/blzcompress.py

The python script does compression when you let a file  be converted into a supported compressed zlib file with a `blz2` header information at the start of the file, the byte orders are in `little`, assume Little Endian, and the compressed file is **Signed**. this compressions works as intended to be



## Compression Error
if file is above UINT16 size limitation (64KB). you'll get this error
![BLZ2Overlapping](https://github.com/user-attachments/assets/c3f46e5e-148c-4d7d-8081-b11b9951cbbc)


## example of modification:
- Original and replacement Texture of Rufus Caligula inside of **"876_Ship_DlcSurvival045.res" > "MS_010C_pr_calig"**
![sample_img](https://github.com/user-attachments/assets/5d43d739-d9c2-449d-a158-7966f28c62ce)
![sample_img2](https://github.com/user-attachments/assets/97c38adf-b1de-4e12-973a-d2573acfe64c)

- both match with the same sizes
![BLZ2Overlapping2](https://github.com/user-attachments/assets/ad277dd3-ae7e-439a-8cf0-49f096a02c6a)


- The reason the file compression fails is that the python script uses a Signed method of compression, so if you use the signed compression to **False**. It will accept that big size file to be compressed.. (image here is just a modified `blzcompress.py`)
![BypassSigned](https://github.com/user-attachments/assets/5e5b0d92-e380-44c6-bee9-7800c271925b)


- but since the original compressed size: 0xAFEA (45034), and the new compressed size is: 0xCD23 (52515)
you will need to heavily reassign the entire structure of **"876_Ship_DlcSurvival045.res"** and **"MS_010C_pr_calig.res"**

**Comment by DarkGD123** [comment]
Any updates so far for new documentation? @nachotacos69

**Reply by nachotacos69** [reply]
Next year I suppose, HaoJun literally mention that

![stare](https://github.com/user-attachments/assets/be94912f-9a56-459b-af53-a91d2e45eaa3)

**Comment by DarkGD123** [comment]
Yo @nachotacos69. if you have the time, can you try to explain or probably recap the following? (would be appreciated)
- TOC (idk how HaoJun calls it on his end). usually the stuff that you had on scripts
- Suggested Audio Quality (at3/at9)
- Compression
- and best way to read the structure

![image](https://github.com/user-attachments/assets/71079b7c-49f4-4132-a226-302d34f2f280)

**Reply by nachotacos69** [reply]
I'll explain whenever i can on my end (we will have different opinions
# TOC
- For the TOC. i usually refer to the table lists that's in within 0x20-0x5F mark. you can pretty much do **jump to the table to 0x20 provided in 0x4, then use [table count or 8 for example]*8=64 or in length it'll be 40 in hex**
![image](https://github.com/user-attachments/assets/160aedcd-9ce0-4111-a9af-26f41aa6b144)
- then you let your script read that 64 bytes of data and tell the script that in that 64 bytes you captured _(start from 0x20 obviously)_, split read it by 8 bytes. So you can organize a list of Groups or Entries than you can assign with.
- Each 8 bytes has it's purpose. first 4 is the offset on where that group is located within that file, the second 4 bytes refers to the count or how many groups are in that area/chunk, start from the offset, then the count is multiplied by 32 bytes. Why?, i dunno, i made it up myself. it's the best way to tell on how many groups are in there easily.

## **But to be honest, you could pretty much see this stuff on the GECV-OLD repo, within the PRES related codes.**

# Audio Quality
- I suggest using 48KHz on AT3 for lower file size while it still atleast give the original audio a good quality. I dunno much on AT9 unless i bought a PSVita and mess around with it.

# Compression
- GEB uses GZip. (with checksum+header information)
- GE2 uses ZLIB.
- it's kinda complicated on how we can compress files with 0xFFFF as the limitation, unless we set the signed compression to be false, the compression works okay. but i think this part still needs better documentation on what compression signature is best used.

# Best way to read the Structure of the .res archive
- Refer to the Documentation HaoJun made to be honest. i mean, i have my documentation too, but i can't really say that i did it right.
-

**Reply by nachotacos69** [reply]
Always check the **GECV-OLD** repo, you can get a bunch of good information there.

**Comment by nachotacos69** [comment]
I tried to make a BLZ compression using C#. I can't really say if it's any good though. Lol. I kinda named it Block Zip, probably becaus of that one blockZip text existing within `EBOOT.BIN`

Release: https://github.com/nachotacos69/GE2Extra/releases/tag/Release

Source Code is in the Repo, you can modify it whenever you like to :D

**Comment by DarkGD123** [comment]
Do you know anything about the PSVita and PS4 versions @nachotacos69

**Comment by nachotacos69** [comment]
RDP Notes:

For each RDP files within the game, (PSVITA doesn't use data anymore and are now using a `data_` folder). the RDP files will have the same signature header structure (EXCEPT FOR DLC in 0x10). each padding is `800` in length, or 2048 bytes.

the game likes to reuse a lot of same offsets within different .res files. so try to use a proper tracker when repacking to avoid making the RDP big because of these stuff

**Comment by nachotacos69** [comment]
game modules notes (sprx/prx/bin/self)

certain texts are not within TR2 or KST files, so HEX/Binary editing is needed to replace those, but if the text count/length is bigger than the original, find a way to optimize your text or make it less or matching with original.
refer to: https://github.com/HaoJun0823/GECV/discussions/24#discussioncomment-11508419

**Comment by nachotacos69** [comment]
Took me time to translate GE Offshot.. most of it are done, @DarkGD123 and the release is on public now.
![2025-04-14-194534-058502](https://github.com/user-attachments/assets/34d630fb-150f-41a8-8b4c-0f4372ebb7d8)
![2025-04-14-194224-885844](https://github.com/user-attachments/assets/dcf4172c-3c97-4266-bad1-59c1ee1fba00)
![2025-04-14-194207-034956](https://github.com/user-attachments/assets/1bb9af17-c8dc-4bf4-8a9c-37c437311a95)
![2025-04-14-194121-940193](https://github.com/user-attachments/assets/4b7ddb88-afb7-4822-864b-3cf9b7deb03b)
![2025-04-13-101924-069120](https://github.com/user-attachments/assets/f018ff67-72ed-40f5-8281-dfe26d7b9e2d)
![2025-04-13-102004-677490](https://github.com/user-attachments/assets/b451fc75-6dd5-4795-bfd3-22c5ebefba70)

**Comment by nachotacos69** [comment]
![model_p1 ixo_Sat_May_31_15-25-06_2025](https://github.com/user-attachments/assets/0e2de675-00bc-42b3-a5d1-d4f89b586b93)
![Event_sub1 ixo_Sat_May_31_14-16-00_2025](https://github.com/user-attachments/assets/27fa7382-3df9-415c-a2bf-fbedce608c80)
I am currently trying to understand IXO model files, so yeah.. while there's any good news on new tools on GECV, I am just doing this for fun..

**Comment by nachotacos69** [comment]
@DarkGD123. btw this is the script that pretty much translates GOD EATER 2 PSP in English. only in 1.30

https://github.com/Lumino5501/PSP-God-Eater-2-English-Patch
the script is originally from @gil-unx. it only focuses translating the text within TR2/KST files, and some images replacement exist but with slight issues but work..
Only the downside is some texts like bullets and other things containing with texts are not editable and the script doesnt handle them at all. which is a poor result but it does its job anyways.

**Comment by nachotacos69** [comment]
@DarkGD123 @GerbaraHenzu

some compression comparisons.. can be useful for future use

i've been tinkering around C++ lately and got some good stuff when compressing data back to raw DEFLATE (RFC 1951).
Python's zlib or zlib in general has some issues with modern processors. so i had to swap to libdeflate (https://github.com/ebiggers/libdeflate). to get the best compressions possible


```
-----------Compression Benchmark Comparisons-----------
=======================================================================================
Input File: tcretab-2-2nd.pgf
Original Size: 1277888 bytes
Iterations per level: 9 for ZLIB, 12 for LIBDEFLATE


// ZLIB raw DEFLATE Compression (Python)
------------------------------------------------------------------------------------------------------------------------
Level       Min (ms)    Avg (ms)    Max (ms)   Compressed Size          Ratio         Saved (%)
------------------------------------------------------------------------------------------------------------------------
1             74.112      79.888     105.608            981823         0.7683            23.17%
2             74.914      78.672      95.945            977437         0.7649            23.51%
3             74.156      82.526     106.587            973541         0.7618            23.82%
4             76.538      80.883      86.986            972833         0.7613            23.87%
5             82.874      85.836      90.014            966785         0.7565            24.35%
6             94.423      96.901     102.940            963670         0.7541            24.59%
7            102.376     104.579     109.561            962909         0.7535            24.65%
8            116.994     120.347     126.436            962389         0.7531            24.69%
9            134.672     137.942     143.689            962386         0.7531            24.69%
------------------------------------------------------------------------------------------------------------------------


// LIBDEFLATE Compression (C++)
------------------------------------------------------------------------------------------------------------------------
Level       Min (ms)    Avg (ms)    Max (ms)   Compressed Size          Ratio         Saved (%)
------------------------------------------------------------------------------------------------------------------------
1             17.853      19.032      21.580            979867         0.7668            23.32%
2             26.773      29.827      40.875            971222         0.7600            24.00%
3             27.356      30.790      33.836            969997         0.7591            24.09%
4             29.174      30.920      36.989            969197         0.7584            24.16%
5             31.066      49.685     102.035            960806         0.7519            24.81%
6             33.367      55.246     107.016            959976         0.7512            24.88%
7             35.547      45.742      63.812            959596         0.7509            24.91%
8             42.153      68.762     206.122            958986         0.7504            24.96%
9             40.727      43.892      50.590            958980         0.7504            24.96%
10            94.727     105.721     119.958            954830         0.7472            25.28%
11           134.062     150.755     190.427            954579         0.7470            25.30%
12           158.846     168.536     201.160            954572         0.7470            25.30%
------------------------------------------------------------------------------------------------------------------------

Summary:

  //Python ZLIB DEFLATE
  Best Compression: Level 9 (962386 bytes, 24.69% saved)
  Fastest: Level 2 (Avg: 78.672 ms)
  
  //C++ LIBDEFLATE
  Best Compression: Level 12 (954572 bytes, 25.30% saved)
  Fastest: Level 1 (Avg: 19.032 ms)
=======================================================================================
```

- the image here shows some changes on the RES file, after some repacking attempts. (.bak are the original ones for the modified RES files)
<img width="618" height="413" alt="image" src="https://github.com/user-attachments/assets/8213b66f-b517-4912-9d5e-a57680a254b0" />

**Comment by nachotacos69** [comment]
Made a python version of the Compression. this should be more better than zlib.
@DarkGD123 

you'll have to install deflate `pip install deflate`
`filename: libdeflate_blz2compress.py`
```
# Replica of RESPlus Compression using libdeflate
# https://github.com/nachotacos69/RESPlus/blob/926ef54eaaddb1ff30c79c1a95b39ec853d1cb1f/RESPlus/PackRES.cpp#L675

import sys
import deflate # https://github.com/dcwatson/deflate
from pathlib import Path

HEADER = b'\x62\x6C\x7A\x32'
MAX_BLOCK_SIZE = 0xFFFF  # 64KB
COMPRESSION_LEVEL = 12 # Set to 12 to compress the data to the smallest level.

def compress_block(data, level):
    # Compress single block and prepend its size as signed 16-bit little-endian.
    compressed = deflate.deflate_compress(data, level)
    size = len(compressed)
    signed_length = size if size < 32768 else size - 65536  # Convert to signed int16
    size_bytes = signed_length.to_bytes(2, byteorder='little', signed=True)
    return size_bytes + compressed

def compress_data(input_data, level=COMPRESSION_LEVEL):
    total_size = len(input_data)
    
    # Single block case: file is MAX_BLOCK_SIZE or smaller
    if total_size <= MAX_BLOCK_SIZE:
        compressed_tail = compress_block(input_data, level)
        return HEADER + compressed_tail
    
    # Multi-block case
    tail_size = total_size % MAX_BLOCK_SIZE
    full_blocks = total_size // MAX_BLOCK_SIZE
    compressed_blocks = []
    
    # Compress tail chunk (head portion)
    if tail_size > 0:
        compressed_blocks.append(compress_block(input_data[:tail_size], level))
    else:
        compressed_blocks.append(b'')
    
    # Compress full 64KB blocks (body)
    for i in range(full_blocks - 1):
        start = tail_size + (i * MAX_BLOCK_SIZE)
        end = start + MAX_BLOCK_SIZE
        compressed_blocks.append(compress_block(input_data[start:end], level))
    
    # Compress last full 64KB block
    if full_blocks > 0:
        start = tail_size + ((full_blocks - 1) * MAX_BLOCK_SIZE)
        compressed_blocks.append(compress_block(input_data[start:], level))
    
    # Rearrange: Header → Last full chunk → Tail chunk → Remaining full chunks
    output = bytearray(HEADER)
    if len(compressed_blocks) > 1:
        output.extend(compressed_blocks[-1])  # Last full chunk
    output.extend(compressed_blocks[0])  # Tail chunk
    for block in compressed_blocks[1:-1]:  # Remaining full chunks
        output.extend(block)
    
    return bytes(output)

def main():
    if len(sys.argv) != 2:
        print(f"Usage: script [input file]")
        sys.exit(1)
    
    input_path = Path(sys.argv[1])
    if not input_path.exists():
        print(f"Error: File '{input_path}' not found")
        sys.exit(1)
    
    # Read input file
    input_data = input_path.read_bytes()
    original_size = len(input_data)
    
    # Compress
    compressed_data = compress_data(input_data)
    compressed_size = len(compressed_data)
    
    # Write output
    output_path = Path.cwd() / f"{input_path.name}.blz2"
    output_path.write_bytes(compressed_data)
    
    # Print statistics
    ratio = (1 - compressed_size / original_size) * 100 if original_size > 0 else 0
    print(f"Original size: {original_size:,} bytes")
    print(f"Compressed size: {compressed_size:,} bytes")
    print(f"Compression ratio: {ratio:.2f}%")
    print(f"Output: {output_path}")

if __name__ == "__main__":
    main()
 ```
